ArduCopter/mode_auto.cpp

```diff
@@ -1162,118 +1178,208 @@ void ModeAuto::nav_attitude_time_run()
             gcs().send_text(MAV_SEVERITY_INFO, "%s landed", prefix_str);
-            nav_payload_place.state = PayloadPlaceStateType_Release;
+            state = State::Release;
             break;
-        case PayloadPlaceStateType_Release:
-        case PayloadPlaceStateType_Releasing:
-        case PayloadPlaceStateType_Delay:
-        case PayloadPlaceStateType_Ascent_Start:
-        case PayloadPlaceStateType_Ascent:
-        case PayloadPlaceStateType_Done:
+        case State::Release:
+        case State::Releasing:
+        case State::Delay:
+        case State::Ascent_Start:
+        case State::Ascent:
+        case State::Done:
             break;
         }
     }
 
 #if AP_GRIPPER_ENABLED == ENABLED
     // if pilot releases load manually:
-    if (g2.gripper.valid() && g2.gripper.released()) {
-        switch (nav_payload_place.state) {
-        case PayloadPlaceStateType_FlyToLocation:
-        case PayloadPlaceStateType_Descent_Start:
-            set_submode(SubMode::NAV_PAYLOAD_PLACE);
+    if (AP::gripper() != nullptr &&
+        AP::gripper()->valid() && AP::gripper()->released()) {
+        switch (state) {
+        case State::Descent_Start:
             gcs().send_text(MAV_SEVERITY_INFO, "%s Manual release", prefix_str);
-            nav_payload_place.state = PayloadPlaceStateType_Done;
+            state = State::Done;
             break;
-        case PayloadPlaceStateType_Descent:
+        case State::Descent_Measure:
+        case State::Descent_Test:
             gcs().send_text(MAV_SEVERITY_INFO, "%s Manual release", prefix_str);
-            nav_payload_place.state = PayloadPlaceStateType_Release;
+            state = State::Release;
             break;
-        case PayloadPlaceStateType_Release:
-        case PayloadPlaceStateType_Releasing:
-        case PayloadPlaceStateType_Delay:
-        case PayloadPlaceStateType_Ascent_Start:
-        case PayloadPlaceStateType_Ascent:
-        case PayloadPlaceStateType_Done:
+        case State::Release:
+        case State::Releasing:
+        case State::Delay:
+        case State::Ascent_Start:
+        case State::Ascent:
+        case State::Done:
             break;
         }
     }
 #endif
 
-    switch (nav_payload_place.state) {
-    case PayloadPlaceStateType_FlyToLocation:
-        if (copter.wp_nav->reached_wp_destination()) {
-            payload_place_start();
+    auto &inertial_nav = copter.inertial_nav;
+    auto &g2 = copter.g2;
+    const auto &g = copter.g;
+    const auto &wp_nav = copter.wp_nav;
+    const auto &pos_control = copter.pos_control;
+
+    if (copter.rangefinder_state.enabled && is_positive(g2.pldp_range_finder_drop_m)) {
+        if (!(copter.rangefinder_alt_ok() && (copter.rangefinder_state.glitch_count == 0) && 
+        (copter.rangefinder_state.alt_cm > g2.pldp_range_finder_drop_m * 100.0 || copter.rangefinder_state.alt_cm < 0.5 * g2.pldp_range_finder_drop_m * 100.0)) ) {
+            rangefinder_drop_alt_time_ms = now_ms;
```

Comment:
```suggestion
    if (!within_rangefinder_release_range()) {
        // reset timer
        rangefinder_drop_alt_time_ms = now_ms;
    } else if (now_ms - rangefinder_drop_alt_time_ms < rangefinder_drop_check_duration_ms) {
```

And add this method:
```
// returns true if rangefinder returns valid range to trigger drop
bool PayloadPlace::within_rangefinder_release_range() const
{
    if (!is_positive(g2.pldp_range_finder_drop_m)) {
        // feature disabled
    }
    if (!copter.rangefinder_state.enabled) {
        // no rangefinder!
        return false;
    }
    if (!copter.rangefinder_alt_ok()) {
        // no valid range from rangefinder
        return false;
    }
    if (copter.rangefinder_state.glitch_count != 0) {
        // do not accept reading when glitching is happening
        return false;
    }

    const float rangefinder_range_m = copter.rangefinder_state.alt_cm * 0.01;
    if (rangefinder_range_m > g2.pldp_range_finder_drop_m) {
        // too high
        return false;
    }

    if (rangefinder_range_m < 0.5 * g2.pldp_range_finder_drop_m) {
        // too low
        return false;
    }

    return true;
}
```

Reply:
done